storage/aws/aws.go

```diff
@@ -565,140 +598,151 @@ func (s *spannerSequencer) assignEntries(ctx context.Context, entries []*tessera
+	}
+	klog.V(1).Infof("Consuming from %d", fromSeq)
 
-			if orderCheck != seq {
-				return fmt.Errorf("integrity fail - expected seq %d, but found %d", orderCheck, seq)
-			}
+	// Now read the sequenced starting at the index we got above.
+	rows, err := tx.QueryContext(ctx, "SELECT seq, v FROM Seq WHERE id = ? AND seq >= ? ORDER BY seq LIMIT ? FOR UPDATE", 0, fromSeq, limit)
+	if err != nil {
+		return false, fmt.Errorf("failed to read Seq: %v", err)
+	}
+	defer rows.Close()
 
-			g := gob.NewDecoder(bytes.NewReader(vGob))
-			b := []storage.SequencedEntry{}
-			if err := g.Decode(&b); err != nil {
-				return fmt.Errorf("failed to deserialise v: %v", err)
-			}
-			entries = append(entries, b...)
-			seqsConsumed = append(seqsConsumed, seq)
-			orderCheck += int64(len(b))
-		}
-		if len(seqsConsumed) == 0 && !forceUpdate {
-			klog.V(1).Info("Found no rows to sequence")
-			return nil
-		}
+	seqsConsumed := []any{}
+	entries := make([]storage.SequencedEntry, 0, limit)
+	orderCheck := fromSeq
+	for rows.Next() {
 
-		// Call consumeFunc with the entries we've found
-		if err := f(ctx, uint64(fromSeq), entries); err != nil {
-			return err
+		var vGob []byte
+		var seq uint64
+		if err := rows.Scan(&seq, &vGob); err != nil {
+			return false, fmt.Errorf("failed to scan Seq row: %v", err)
 		}
 
-		// consumeFunc was successful, so we can update our coordination row, and delete the row(s) for
-		// the then consumed entries.
-		m := make([]*spanner.Mutation, 0)
-		m = append(m, spanner.Update("IntCoord", []string{"id", "seq"}, []interface{}{0, int64(orderCheck)}))
-		for _, c := range seqsConsumed {
-			m = append(m, spanner.Delete("Seq", spanner.Key{0, c}))
+		if orderCheck != seq {
+			return false, fmt.Errorf("integrity fail - expected seq %d, but found %d", orderCheck, seq)
 		}
-		if len(m) > 0 {
-			if err := txn.BufferWrite(m); err != nil {
-				return err
-			}
+
+		g := gob.NewDecoder(bytes.NewReader(vGob))
+		b := []storage.SequencedEntry{}
+		if err := g.Decode(&b); err != nil {
+			return false, fmt.Errorf("failed to deserialise v from Seq: %v", err)
 		}
+		entries = append(entries, b...)
+		seqsConsumed = append(seqsConsumed, seq)
+		orderCheck += uint64(len(b))
+	}
+	if len(seqsConsumed) == 0 && !forceUpdate {
+		klog.V(1).Info("Found no rows to sequence")
+		return false, nil
+	}
 
-		didWork = true
-		return nil
-	})
-	if err != nil {
+	// Call consumeFunc with the entries we've found
+	if err := f(ctx, uint64(fromSeq), entries); err != nil {
 		return false, err
 	}
 
-	return didWork, nil
+	// consumeFunc was successful, so we can update our coordination row, and delete the row(s) for
+	// the then consumed entries.
+	if _, err := tx.ExecContext(ctx, "UPDATE IntCoord SET seq=? WHERE id=?", orderCheck, 0); err != nil {
+		return false, fmt.Errorf("update intcoord: %v", err)
+	}
+
+	if len(seqsConsumed) > 0 {
+		q := "DELETE FROM Seq WHERE id=? AND seq IN ( " + placeholder(len(seqsConsumed)) + " )"
```

Comment:
Not for now, but worth looking at seeing how `seq IN (?, ?, ? ...)` performs vs, say, `seq BETWEEN ? AND ?` - in theory the latter _may_ be faster as it can do it in a single range scan vs. potentially having to do a sequential bunch of "pecks", but would need to be looked into once you have it all working.